test/test_autograd.py

```diff
@@ -600,6 +600,35 @@ def test_grad_unreachable(self):
                                     "Set allow_unused=True"):
             grad_x, grad_y = torch.autograd.grad(x * 2, [x, y], allow_unused=False)
 
+    def test_grad_unreachable_discovery(self):
+        # Test that certain nodes are not erroneously executed when an input
+        # is unreachable. See #39784
+        class MyFunc(torch.autograd.Function):
+            @staticmethod
+            def forward(ctx, x):
+                return x
+
+            @staticmethod
+            def backward(ctx, x):
+                self.fail("This node should not be executed!")
+
+        x = MyFunc.apply(torch.randn(1, requires_grad=True) * 2)
+        y = torch.randn(1, requires_grad=True)
+        (gY,) = torch.autograd.grad(x, (y, ), allow_unused=True)
+        self.assertIsNone(gY)
+
+        x = MyFunc.apply(torch.randn(1, requires_grad=True) * 2)
+        y = torch.randn(1, requires_grad=True)
+        z = torch.randn(1, requires_grad=True)
+        (gY, gZ) = torch.autograd.grad(x + z, (y, z), allow_unused=True)
+        self.assertIsNone(gY)
+        self.assertIsNotNone(gZ)
+
+        x = MyFunc.apply(torch.randn(1, requires_grad=True) * 2)
+        y = torch.randn(1, requires_grad=True)
+        torch.autograd.backward(x, inputs=(y, ))  # allow_unused is implicitly True!
```

Comment:
You mean False?

Reply:
It is False by default for .grad() but seems to be True for .backward() since its does not raise an error here

Reply:
Ho right sorry. I though this was a call to autograd.grad because you are mentioning this flag.
It does not exist for backward, so yes it is as if it was always True.

Reply:
Would it also make sense to have something like "allow_unused" for backward for symmetry? I guess its going to be tough making the default False at this point bc it would be bc breaking, but seems like by default unused inputs should raise an inputs for backward as well.

Reply:
I am not sure we want to add that flag.
Also it might be tricky to implement because we don't capture anything in this case and so it will be hard to know if the Nodes were used or not.